harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -29,14 +29,70 @@
     pass
 
 
+class TrainUnit:
+    def __init__(self, value):
+        self.value = value
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level workload sequencer loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+    pass
+
+
+class SavableState:
+    def __init__(
+        self,
+        trial_id: int,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
+        step_id: int = 0,
+        last_val: int = 0,
+    ) -> None:
+        # Store TrialID to distinguish between e.g. pause/restart and continue training.
+        self.trial_id = trial_id
+        self.last_ckpt = last_ckpt
+        self.steps_completed = steps_completed
+        self.step_id = step_id
+        self.last_val = last_val
+
+
 class PyTorchTrialController(det.TrialController):
     def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
 
         check.is_instance(trial_inst, PyTorchTrial, "PyTorchTrialController needs an PyTorchTrial")
         self.trial = cast(PyTorchTrial, trial_inst)
         self.context = cast(pytorch.PyTorchTrialContext, self.context)
+        self.core_context = self.context._core
         self.context._set_determined_profiler(self.prof)
+
+        # Initialize training loop variables
+        self.max_length = None
+        self.checkpoint_period = None
+        self.validation_period = None
+        self.batches_trained = 0
+        self.epochs_trained = 0
+        self.training_metrics = []
+        self.searcher_op = None
+
+        self.records_per_epoch = self.context.env.experiment_config.get_records_per_epoch()
+        self._trial_id = int(self.env.det_trial_id)
```

Comment:
why is this prefixed with an underscore but nothing else is?

Reply:
i'm gonna do a round at end of reviews to clean up everything like this.

Reply:
Why not now?

```
git commit -am "going to try sed"
sed -e 's/self._trial_id/self.trial_id/g' _pytorch_trial.py > temp
mv temp _pytorch_trial.py
```

If you have an editor with multiple cursor support, this could be done interactively in like 10 seconds.

(my suggestion is for your sake, not mine.  The more stuff I keep on a mental list of "oh yeah I need to do that later" the more stressed out I get)